mmdet/models/dense_heads/maskformer_head.py

```diff
@@ -0,0 +1,914 @@
+                ignored. Defaults to None.
+
+        Returns:
+            losses (dict[str, Tensor]): a dictionary of loss components
+        """
+        assert gt_bboxes_ignore is None  # not consider ignoring bboxes
+
+        # forward
+        all_cls_scores, all_mask_preds = self(feats)
+
+        # preprocess ground truth
+        gt_labels, gt_masks = self.preprocess_ground_truth(
+            gt_labels, gt_masks, gt_semantic_seg)
+
+        # loss
+        losses = self.loss(all_cls_scores, all_mask_preds, gt_labels, gt_masks,
+                           img_metas)
+
+        return losses
+
+    def simple_test(self, feats, img_metas, rescale=False):
+        """Test segment without test-time aumengtation.
+
+        Only the output of last decoder layers was used.
+
+        Args:
+            feats (list[Tensor]): Multi-level features from the
+                upstream network, each is a 4D-tensor.
+            img_metas (list[dict]): List of image information.
+            rescale (bool, optional):  If True, return boxes in
+                original image space. Default False.
+
+        Returns:
+            list[dict[str, np.array]]: semantic segmentation results
+                and panoptic segmentation results for each image.
+                [
+                    {
+                        'pan_results': np.array, # shape = [h, w]
+                    },
+                    ...
+                ]
+        """
+        all_cls_scores, all_mask_preds = self(feats)
+        mask_cls_results = all_cls_scores[-1]
+        mask_pred_results = all_mask_preds[-1]
+
+        # upsample masks
+        img_shape = img_metas[0]['pad_shape'][:2]
+        mask_pred_results = F.interpolate(
+            mask_pred_results,
+            size=(img_shape[0], img_shape[1]),
+            mode='bilinear',
+            align_corners=False)
+
+        results = []
+        for mask_cls_result, mask_pred_result, meta in zip(
+                mask_cls_results, mask_pred_results, img_metas):
+            # remove padding
+            img_height, img_width = meta['img_shape'][:2]
+            mask_pred_result = mask_pred_result[:, :img_height, :img_width]
+
+            if rescale:
+                # return result in original resolution
+                ori_height, ori_width = meta['ori_shape'][:2]
+                mask_pred_result = F.interpolate(mask_pred_result.unsqueeze(1),
+                                                 size=(ori_height, ori_width),
+                                                 mode='bilinear',
+                                                 align_corners=False)\
+                    .squeeze(1)
+
+            mask = self.post_process(mask_cls_result, mask_pred_result)
+            result = {'pan_results': mask.detach().cpu().numpy()}
+            results.append(result)
+
+        return results
+
+    def post_process(self, mask_cls, mask_pred):
+        """Panoptic segmengation inference.
```

Comment:
Should add reference as this code is borrowed from MaskFormer. See example here https://github.com/open-mmlab/mmdetection/blob/master/mmdet/models/roi_heads/mask_heads/fcn_mask_head.py#L347